### PR TITLE
feat: Support multi-epoch training in SFT

### LINKED ISSUE #143: Support multi-epoch training with SFT

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

**Add a one line overview of what this PR aims to accomplish.**

# Issues
closes #143 

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA/reinforcer/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA/reinforcer/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA/reinforcer/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

File: nemo_reinforcer/algorithms/sft.py [hunk 1]
```diff
@@ -334,124 +345,144 @@ def sft_train(
...
+                            "input_lengths": input_lengths,
+                            "token_mask": cat_and_padded["token_loss_mask"],
+                            "sample_mask": batch["loss_multiplier"],
+                        }
                     )
-                    torch.save(
-                        train_dataloader.state_dict(),
-                        os.path.join(checkpoint_path, "train_dataloader.pt"),
+
+                ## train_data.to("cpu")
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/sft.py]: debug comment?

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 2]
```diff
@@ -955,7 +955,7 @@ def train(
         # Shard and replicate the batch
         shards = self.dp_size
         sharded_data = data.shard_by_batch_size(
-            shards, batch_size=self.cfg["train_global_batch_size"]
+            shards, batch_size=gbs or self.cfg["train_global_batch_size"]
```
Comment (terrykong) [inline nemo_reinforcer/models/policy/hf_policy.py]: can you mirror this logic on line 968? It doesn't change anything, but the it's more readable and we don't implicitly rely on the actor to do the right thing

File: nemo_reinforcer/algorithms/sft.py [hunk 3]
```diff
@@ -334,124 +345,144 @@ def sft_train(
...
+                    print(f"  • {k}: {v:.2f}s ({percent:.1f}%)")
+
+            logger.log_metrics(metrics, total_steps + 1, prefix="train")
+            logger.log_metrics(timing_metrics, total_steps + 1, prefix="timing/train")
+
+            timer.reset()
+            current_step += 1
+            total_steps += 1
+
+            if total_steps >= master_config["sft"]["max_num_steps"]:
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/sft.py:520]: WDYT about moving this to the while loop that checks for epochs? Just so we're consistent that the nothing will happen if max epochs or max steps is reached

File: examples/configs/sft.yaml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 # SFT Algorithm Configuration
 sft:
```
Comment (terrykong) [inline examples/configs/sft.yaml:2]: Could you add a comment explaining the behavior here that the "smallest" of these two will take affect first

File: nemo_reinforcer/algorithms/sft.py (hunk 3 above)
Comment (terrykong) [reply nemo_reinforcer/algorithms/sft.py:520]: actually, it makes sense to keep this logic here, but maybe also copy it on the while loop line

File: nemo_reinforcer/algorithms/sft.py [hunk 5]
```diff
@@ -334,124 +345,144 @@ def sft_train(
...
-                sft_save_state["step"] = step + 1
-                sft_save_state["val_loss"] = val_metrics["val_loss"]
-                with timer.time("checkpointing"):
-                    print(f"Saving checkpoint for step {step + 1}...")
-                    checkpoint_path = checkpointer.init_tmp_checkpoint(
-                        step + 1, sft_save_state, master_config
+    while current_epoch < max_num_epochs:
+        print(f"\n{'=' * 25} Epoch {current_epoch + 1}/{max_num_epochs} {'=' * 25}")
+
+        for batch in train_dataloader:
```
Comment (terrykong) [inline nemo_reinforcer/algorithms/sft.py:383]: is it possible to add a unit test for this? maybe pass an empty data loader and do a print right before the while loop and make sure the last line [captured](https://docs.pytest.org/en/7.1.x/how-to/capture-stdout-stderr.html#accessing-captured-output-from-a-test-function) is that added print and not "Epoch {current_epoch}.... "

Comment (ashors1) [reply nemo_reinforcer/algorithms/sft.py:383]: Can you check whether the unit tests I just added are sufficient?
